### ISSUE TITLE
Algorithm 2 inconsistent with code

### ISSUE THREAD

**Issue by tianboh** [litian96/FedProx#22]
Thanks for the work :) 
I have read the code, and corresponding issues #10, but there are some places I still feel inconsistent with the paper. Please correct me if I am wrong.
- In the Algorithm 2 line 7, we are calculating the norm between local model and global model. But the code is using l2 norm for local model without considering global model. Taks [mnist/mclr.py line 40](https://github.com/litian96/FedProx/blob/master/flearn/models/mnist/mclr.py#L40) for example.

- I also checked the NLP experiments Shakespare, but I didn't find the regularization part in create_model. [shakespare/stacked_lstm.py create_model](https://github.com/litian96/FedProx/blob/master/flearn/models/shakespeare/stacked_lstm.py#L51)

Thank you!

![fig1](https://user-images.githubusercontent.com/54729592/114845202-29073100-9e0e-11eb-81fa-e0a435ff411b.png)

**Comment by litian96** [comment]
The updates of FedProx is defined here: https://github.com/litian96/FedProx/blob/master/flearn/optimizer/pgd.py#L29-L36
where var is the current local model you are optimizing for, and vstar is the current global model which is the starting point of solving for the local subproblem.

**Comment by tianboh** [comment]
Thank you, It makes sense now!

**Comment by tianboh** [comment]
Hi Tian, 

Thanks for previous reply, but there is one place I still don't figure out. The paper shows objective function is optimizing Fk(w) and ||w - w^t|| simultaneously, but the code shows that we are calculating loss and ||w - w^t|| seperately. I found issue #19 #18 and #6 are quite useful, but not exactly what I need.

Take mnist task as an example, the code first calculate the cross entropy loss in 
[grads_and_vars = optimizer.compute_gradients(loss)](https://github.com/litian96/FedProx/blob/master/flearn/models/mnist/mclr.py#L47), this part does not consider ||w - w^t||. 
By calling [var_update = state_ops.assign_sub(var, lr_t*(grad + mu_t*(var-vstar)))](https://github.com/litian96/FedProx/blob/master/flearn/models/mnist/mclr.py#L47) in 
[train_op = optimizer.apply_gradients(grads_and_vars, global_step=tf.train.get_global_step())](https://github.com/litian96/FedProx/blob/master/flearn/models/mnist/mclr.py#L49) we consider the proximal term.

```python
  def create_model(self, optimizer):
      """Model function for Logistic Regression."""
      features = tf.placeholder(tf.float32, shape=[None, 784], name='features')
      labels = tf.placeholder(tf.int64, shape=[None,], name='labels')
      logits = tf.layers.dense(inputs=features, units=self.num_classes, kernel_regularizer=tf.contrib.layers.l2_regularizer(0.001))
      predictions = {
          "classes": tf.argmax(input=logits, axis=1),
          "probabilities": tf.nn.softmax(logits, name="softmax_tensor")
          }
      loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
      # set_trace()
      grads_and_vars = optimizer.compute_gradients(loss)
      grads, _ = zip(*grads_and_vars)
      train_op = optimizer.apply_gradients(grads_and_vars, global_step=tf.train.get_global_step())
      eval_metric_ops = tf.count_nonzero(tf.equal(labels, predictions["classes"]))
      return features, labels, train_op, grads, eval_metric_ops, loss
```


Therefore, it seems that the code is calculating the gradient from Fk's perspective, and then subtract a proximal part to get the final gradient. I also debuged the code to check who has called pgd's [_apply_dense()](https://github.com/litian96/FedProx/blob/master/flearn/optimizer/pgd.py#L29), it shows that this method is only used by optimizer.apply_gradient(), instead of compute_gradient().

Thank you in advance :)

**Comment by litian96** [comment]
You are exactly right that we first need to obtain the gradients of \nabla F_k (w_local) + \mu (w_local - w_global), and then apply the gradients to w_local.
What is your question?

**Comment by tianboh** [comment]
My question is compute_gradient() only calculates \nabla F_k(w_local). How can we find argmin_w h_k without considering \nabla (w_local - w_global)? 

I know we considered \mu (w_local - w_global) in apply_gradient(), but this proximal term not necessary leads to argmin h_k.

**Comment by litian96** [comment]
The gradient of h_k is \nabla F_k (w_local) + \mu (w_local - w_global), which is what we apply to update w_local (specifically, `grad + mu_t * (var-vstar) `).

**Comment by tianboh** [comment]
Ohhhhh, I see, really appreciate for the clarification :)
